addons/nuxeo-core-binarymanager-cloud/nuxeo-core-binarymanager-s3/src/main/java/org/nuxeo/ecm/blob/s3/S3BlobStore.java

```diff
@@ -0,0 +1,782 @@
+                if (!found) {
+                    return false;
+                }
+                file = tmp;
+                fileTraceSource = "tmp";
+            }
+            String versionId = writeFile(key, file, null, fileTraceSource); // always atomic
+            if (versionId != null) {
+                throw new NuxeoException("Cannot copy blob if store has versioning");
+            }
+            if (atomicMove) {
+                sourceStore.deleteBlob(sourceKey);
+            }
+            return true;
+        } finally {
+            if (tmp != null) {
+                try {
+                    logTrace("tmp", "-->", "tmp", "delete");
+                    logTrace("hnote right: " + tmp.getFileName());
+                    Files.delete(tmp);
+                } catch (IOException e) {
+                    log.warn(e, e);
+                }
+            }
+        }
+    }
+
+    @Override
+    public void writeBlobProperties(BlobUpdateContext blobUpdateContext) throws IOException {
+        String key = blobUpdateContext.key;
+        String objectKey;
+        String versionId;
+        int seppos = key.indexOf(VER_SEP);
+        if (seppos < 0) {
+            objectKey = key;
+            versionId = null;
+        } else {
+            objectKey = key.substring(0, seppos);
+            versionId = key.substring(seppos + 1);
+        }
+        String bucketKey = bucketPrefix + objectKey;
+        try {
+            if (blobUpdateContext.updateRetainUntil != null) {
+                if (versionId == null) {
+                    throw new IOException("Cannot set retention on non-versioned blob");
+                }
+                Calendar retainUntil = blobUpdateContext.updateRetainUntil.retainUntil;
+                Date retainUntilDate = retainUntil == null ? null : retainUntil.getTime();
+                ObjectLockRetention retention = new ObjectLockRetention();
+                retention.withMode(objectLockRetentionMode) //
+                         .withRetainUntilDate(retainUntilDate);
+                SetObjectRetentionRequest request = new SetObjectRetentionRequest();
+                request.withBucketName(bucketName) //
+                       .withKey(bucketKey)
+                       .withVersionId(versionId)
+                       .withRetention(retention);
+                logTrace("->", "setObjectRetention");
+                logTrace("hnote right: " + bucketKey + "v=" + versionId);
+                logTrace("rnote right: " + (retainUntil == null ? "null" : retainUntil.toInstant().toString()));
+                amazonS3.setObjectRetention(request);
+            }
+            if (blobUpdateContext.updateLegalHold != null) {
+                if (versionId == null) {
+                    throw new IOException("Cannot set legal hold on non-versioned blob");
+                }
+                boolean hold = blobUpdateContext.updateLegalHold.hold;
```

Comment:
you explained it during our meeting but i forgot the explanation  (sorry)about the fact why we need a versioning to be able to use `retention` and `legal hold`. actually you added a low level check to avoid any remove of the blob during retention. but with versioning are us allowing updating of the main blob

Reply:
It's a S3 requirement: when you activate retention in S3 it automatically activates versioning.